File: alica_engine/src/engine/modelmanagement/factories/VariableBindingFactory.cpp
#include "engine/modelmanagement/factories/VariableBindingFactory.h"
#include "engine/modelmanagement/Strings.h"

namespace alica
{
VariableBinding* VariableBindingFactory::create(const YAML::Node& node)
{
    auto* variableBinding = new VariableBinding();
    Factory::setAttributes(node, variableBinding);
    Factory::storeElement(variableBinding, alica::Strings::variableBinding);

    if (Factory::isValid(node[alica::Strings::variable])) {
        Factory::bindingVarReferences.push_back(
                std::pair<int64_t, int64_t>(variableBinding->getId(), Factory::getReferencedId(node[alica::Strings::variable])));
    }

    if (Factory::isValid(node[alica::Strings::subPlan])) {
        Factory::bindingSubPlanReferences.push_back(
                std::pair<int64_t, int64_t>(variableBinding->getId(), Factory::getReferencedId(node[alica::Strings::subPlan])));
    }

    if (Factory::isValid(node[alica::Strings::subVariable])) {
        Factory::bindingSubVarReferences.push_back(
                std::pair<int64_t, int64_t>(variableBinding->getId(), Factory::getReferencedId(node[alica::Strings::subVariable])));
    }

    return variableBinding;
}

void VariableBindingFactory::attachReferences()
{
    for (std::pair<int64_t, int64_t> pairs : Factory::bindingSubPlanReferences) {
        VariableBinding* p = (VariableBinding*) Factory::getElement(pairs.first);
        AbstractPlan* ap = (AbstractPlan*) Factory::getElement(pairs.second);
        p->setSubPlan(ap);
    }
    Factory::bindingSubPlanReferences.clear();

    for (std::pair<int64_t, int64_t> pairs : Factory::bindingSubVarReferences) {
        VariableBinding* p = (VariableBinding*) Factory::getElement(pairs.first);
        Variable* ap = (Variable*) Factory::getElement(pairs.second);
        p->setSubVar(ap);
    }
    Factory::bindingSubVarReferences.clear();

    for (std::pair<int64_t, int64_t> pairs : Factory::bindingVarReferences) {
        VariableBinding* p = (VariableBinding*) Factory::getElement(pairs.first);
        Variable* v = (Variable*) Factory::getElement(pairs.second);
        p->setVar(v);
    }
    Factory::bindingVarReferences.clear();
}
} // namespace alica

File: alica-supplementary/plan-conversion-tool/src/main/java/com/rapyutarobotics/alica/factories/VariableBindingFactory.java
package com.rapyutarobotics.alica.factories;

import com.rapyutarobotics.alica.ConversionProcess;
import com.rapyutarobotics.alica.Tags;
import de.unikassel.vs.alica.planDesigner.alicamodel.Plan;
import de.unikassel.vs.alica.planDesigner.alicamodel.Variable;
import de.unikassel.vs.alica.planDesigner.alicamodel.VariableBinding;
import javafx.util.Pair;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

public class VariableBindingFactory extends Factory {
    public static VariableBinding create (Element variableBindingNode, ConversionProcess cp) {
        VariableBinding variableBinding = new VariableBinding();
        Factory.setAttributes(variableBindingNode, variableBinding);
        cp.addElement(variableBinding);
        NodeList variableNodeList = variableBindingNode.getElementsByTagName(Tags.VARIABLE);
        if (variableNodeList.getLength() > 0) {
            cp.bindingVarReferences.put(variableBinding.getId(), cp.getReferencedId(variableNodeList.item(0).getTextContent()));
        }
        NodeList subPlanNodeList = variableBindingNode.getElementsByTagName(Tags.SUBPLAN);
        if (subPlanNodeList.getLength() > 0) {
            cp.bindingSubPlanReferences.put(variableBinding.getId(), cp.getReferencedId(variableNodeList.item(0).getTextContent()));
        }
        NodeList subVariableList = variableBindingNode.getElementsByTagName(Tags.SUBVARIABLE);
        if (subVariableList.getLength() > 0) {
            cp.bindingSubVarReferences.put(variableBinding.getId(), cp.getReferencedId(variableNodeList.item(0).getTextContent()));
        }
        return variableBinding;
    }

    public static void attachReferences(ConversionProcess cp) {
        for (Pair<Long, Long> entry : cp.bindingSubPlanReferences.getEntries()) {
            VariableBinding variableBinding = (VariableBinding) cp.getElement(entry.getKey());
            Plan plan = (Plan) cp.getElement(entry.getValue());
            if (plan == null) {
                throw new RuntimeException("[VariableBindingFactory] Sub plan with ID " + entry.getValue() + " unknown!");
            }
            variableBinding.setSubPlan(plan);
        }
        cp.bindingSubPlanReferences.clear();

        for (Pair<Long, Long> entry : cp.bindingSubVarReferences.getEntries()) {
            VariableBinding variableBinding = (VariableBinding) cp.getElement(entry.getKey());
            Variable variable = (Variable) cp.getElement(entry.getValue());
            if (variable == null) {
                throw new RuntimeException("[VariableBindingFactory] Sub variable with ID " + entry.getValue() + " unknown!");
            }
            variableBinding.setSubVariable(variable);
        }
        cp.bindingSubVarReferences.clear();

        for (Pair<Long, Long> entry : cp.bindingVarReferences.getEntries()) {
            VariableBinding variableBinding = (VariableBinding) cp.getElement(entry.getKey());
            Variable variable = (Variable) cp.getElement(entry.getValue());
            if (variable == null) {
                throw new RuntimeException("[VariableBindingFactory] Variable with ID " + entry.getValue() + " unknown!");
            }
            variableBinding.setVariable(variable);
        }
        cp.bindingVarReferences.clear();
    }
}
